### PR TITLE
Fix and add Tests For character length and concat

### PR DESCRIPTION
**Pull request by vvidlearn**
#240 #241

### CODE REVIEW

**Comment by vvidlearn** [comment]
hi @jczuchnowski 
Thanks for the nice zio-sql intro talk. 
I was just trying to do my first PR here. Sorry I am very new to scala, so please bear with me on this one.

The current PR is only doing this basic test query for char length `val query = select(CharLength("first_name")) from customers`

but looking at your talk I realised this should be possible `val query = select(CharLength(Concat(fName ++ lName))) from customers`.  

```
 type mismatch;
[error]  found   : zio.sql.postgresql.PostgresModuleTest.Selection[zio.sql.postgresql.PostgresModuleTest.Features.Union[zio.sql.postgresql.PostgresModuleTest.Features.Source,zio.sql.postgresql.PostgresModuleTest.Features.Source],zio.sql.postgresql.PostgresModuleTest.Customers.customers.TableType,zio.sql.postgresql.PostgresModuleTest.SelectionSet.Cons[zio.sql.postgresql.PostgresModuleTest.Customers.customers.TableType,String,zio.sql.postgresql.PostgresModuleTest.SelectionSet.Cons[zio.sql.postgresql.PostgresModuleTest.Customers.customers.TableType,String,zio.sql.postgresql.PostgresModuleTest.SelectionSet.Empty]]]
[error]     (which expands to)  zio.sql.postgresql.PostgresModuleTest.Selection[zio.sql.postgresql.PostgresModuleTest.Features.Union[zio.sql.postgresql.PostgresModuleTest.Features.Source,zio.sql.postgresql.PostgresModuleTest.Features.Source],zio.sql.postgresql.PostgresModuleTest.Customers.customers.TableType,zio.sql.postgresql.PostgresModuleTest.SelectionSet.Cons[zio.sql.postgresql.PostgresModuleTest.Customers.customers.TableType,String,zio.sql.postgresql.PostgresModuleTest.SelectionSet.Cons[zio.sql.postgresql.PostgresModuleTest.Customers.customers.TableType,String,zio.sql.postgresql.PostgresModuleTest.SelectionSet.Empty.type]]]
[error]  required: zio.sql.postgresql.PostgresModuleTest.Expr[?,?,(String, String)]
[error]       val query = select(CharLength(Concat(fName ++ lName))) from customers
```

I would like to add test for a query like `select(Concat(fName ++ lName)) ++ (CharLength(Concat(fName ++ lName))) from customers`. Just wanted to know if its currently possible. thank you.

**Comment by pjazdzewski1990** [comment]
Hi @vvidlearn 
Not sure if it helps, but I was working on something similar (concat_ws) on https://github.com/zio/zio-sql/pull/267 Spoiler alert it's still VERY dirty :) but maybe it will help you
From the stack I guess, that you are constructing a Selection (++) but using it as an Expr. IMO use a comma instead of ++
Having said that, we  both might have issue with extending our solution... We are unlucky to have varargs and atm natively zio-sql supports up to 4 args. I've tried adding varargs yesterday, got stuck in "type hell"

**Comment by vvidlearn** [comment]
Hi @pjazdzewski1990 thanks for the reply and for the generous comment :) 
Watching the talk yesterday I heard about dependent types and  that zio-sql uses it for compile time safely. Hence I was trying to writing a test that I think would combine SelectionSet with FunctionDef but ye the rather lengthy types takes some getting used to. Good luck with your PR too :)

File: postgres/src/test/scala/zio/sql/postgresql/PostgresModuleTest.scala [hunk 1]
```diff
@@ -183,8 +183,41 @@ object PostgresModuleTest extends PostgresRunnableSpec with ShopSchema {
         r <- result.runCollect
       } yield assert(r)(hasSameElementsDistinct(expected))
 
+      assertion.mapErrorCause(cause => Cause.stackless(cause.untraced))
+    },
+    testM("Can concat strings with concat function") {
+
+      import FunctionDef.Concat
+      val query = (select(Concat("first_name", "last_name") as "fullname") from customers)
```
Comment (jczuchnowski) [inline postgres/src/test/scala/zio/sql/postgresql/PostgresModuleTest.scala:191]: No need the outer parentheses

**Comment by jczuchnowski** [comment]
@vvidlearn would you mind moving the tests to FunctionDefSpec.scala? We're keeping these function tests in there.

**Comment by jczuchnowski** [comment]
@vvidlearn to answer your question - no, the syntax of `select(CharLength(Concat(fName ++ lName)))` won't work at this point. I'll create the issue for that.

**Comment by vvidlearn** [comment]
sure will do, thanks

**Comment by vvidlearn** [comment]
@jczuchnowski moved tests to FunctionDefSpec as suggested.

**Review by jczuchnowski — APPROVED** [review]

**Pull request merged**